File: zebra-py/tests/test_engine.py
"""Tests for the workflow engine."""

import pytest

from zebra.core.engine import WorkflowEngine
from zebra.core.exceptions import InvalidStateTransitionError, ProcessNotFoundError
from zebra.core.models import (
    ProcessDefinition,
    ProcessState,
    RoutingDefinition,
    TaskDefinition,
    TaskInstance,
    TaskResult,
    TaskState,
)
from zebra.storage.memory import InMemoryStore
from zebra.tasks.base import ExecutionContext, TaskAction
from zebra.tasks.registry import ActionRegistry


class CountingAction(TaskAction):
    """A test action that counts executions."""

    execution_count = 0

    async def run(self, task: TaskInstance, context: ExecutionContext) -> TaskResult:
        CountingAction.execution_count += 1
        return TaskResult.ok(output={"count": CountingAction.execution_count})


class FailingAction(TaskAction):
    """A test action that always fails."""

    async def run(self, task: TaskInstance, context: ExecutionContext) -> TaskResult:
        return TaskResult.fail("Intentional failure")


@pytest.fixture
def store():
    return InMemoryStore()


@pytest.fixture
def registry():
    reg = ActionRegistry()
    reg.register_action("counting", CountingAction)
    reg.register_action("failing", FailingAction)
    return reg


@pytest.fixture
def engine(store, registry):
    return WorkflowEngine(store, registry)


@pytest.fixture
def simple_definition():
    """A simple two-task sequential workflow."""
    return ProcessDefinition(
        id="simple",
        name="Simple Workflow",
        first_task_id="start",
        tasks={
            "start": TaskDefinition(id="start", name="Start", action="counting"),
            "end": TaskDefinition(id="end", name="End", action="counting"),
        },
        routings=[
            RoutingDefinition(id="r1", source_task_id="start", dest_task_id="end"),
        ],
    )


@pytest.fixture
def parallel_definition():
    """A workflow with parallel branches."""
    return ProcessDefinition(
        id="parallel",
        name="Parallel Workflow",
        first_task_id="start",
        tasks={
            "start": TaskDefinition(id="start", name="Start"),
            "branch_a": TaskDefinition(id="branch_a", name="Branch A", action="counting"),
            "branch_b": TaskDefinition(id="branch_b", name="Branch B", action="counting"),
            "join": TaskDefinition(id="join", name="Join", synchronized=True, action="counting"),
        },
        routings=[
            RoutingDefinition(
                id="r1", source_task_id="start", dest_task_id="branch_a", parallel=True
            ),
            RoutingDefinition(
                id="r2", source_task_id="start", dest_task_id="branch_b", parallel=True
            ),
            RoutingDefinition(id="r3", source_task_id="branch_a", dest_task_id="join"),
            RoutingDefinition(id="r4", source_task_id="branch_b", dest_task_id="join"),
        ],
    )


class TestWorkflowEngine:
    @pytest.mark.asyncio
    async def test_create_process(self, engine, simple_definition):
        process = await engine.create_process(simple_definition)

        assert process.state == ProcessState.CREATED
        assert process.definition_id == simple_definition.id

        # Definition should be saved
        loaded = await engine.store.load_definition(simple_definition.id)
        assert loaded is not None

    @pytest.mark.asyncio
    async def test_start_process(self, engine, simple_definition):
        CountingAction.execution_count = 0

        process = await engine.create_process(simple_definition)
        process = await engine.start_process(process.id)

        assert process.state == ProcessState.COMPLETE
        # Both tasks should have run
        assert CountingAction.execution_count == 2

    @pytest.mark.asyncio
    async def test_process_with_properties(self, engine, simple_definition):
        process = await engine.create_process(simple_definition, properties={"key": "value"})

        assert process.properties["key"] == "value"

    @pytest.mark.asyncio
    async def test_manual_task(self, engine, store, registry):
        """Test a workflow with a manual task."""
        definition = ProcessDefinition(
            id="manual",
            name="Manual Workflow",
            first_task_id="manual_task",
            tasks={
                "manual_task": TaskDefinition(id="manual_task", name="Manual Task", auto=False),
                "end": TaskDefinition(id="end", name="End"),
            },
            routings=[
                RoutingDefinition(id="r1", source_task_id="manual_task", dest_task_id="end"),
            ],
        )

        process = await engine.create_process(definition)
        await engine.start_process(process.id)

        # Process should be running, waiting for manual task
        process = await store.load_process(process.id)
        assert process.state == ProcessState.RUNNING

        # Get pending tasks
        pending = await engine.get_pending_tasks(process.id)
        assert len(pending) == 1
        assert pending[0].task_definition_id == "manual_task"

        # Complete the manual task
        await engine.complete_task(pending[0].id, TaskResult.ok(output="done"))

        # Process should be complete now
        process = await store.load_process(process.id)
        assert process.state == ProcessState.COMPLETE

    @pytest.mark.asyncio
    async def test_parallel_execution(self, engine, parallel_definition):
        """Test parallel branch execution with sync/join."""
        CountingAction.execution_count = 0

        process = await engine.create_process(parallel_definition)
        await engine.start_process(process.id)

        process = await engine.store.load_process(process.id)
        assert process.state == ProcessState.COMPLETE

        # Should have executed: branch_a, branch_b, join
        assert CountingAction.execution_count == 3

    @pytest.mark.asyncio
    async def test_pause_resume(self, engine, store, registry):
        """Test pausing and resuming a workflow."""
        definition = ProcessDefinition(
            id="pausable",
            name="Pausable",
            first_task_id="task1",
            tasks={
                "task1": TaskDefinition(id="task1", name="Task 1", auto=False),
                "task2": TaskDefinition(id="task2", name="Task 2"),
            },
            routings=[
                RoutingDefinition(id="r1", source_task_id="task1", dest_task_id="task2"),
            ],
        )

        process = await engine.create_process(definition)
        await engine.start_process(process.id)

        # Pause
        process = await engine.pause_process(process.id)
        assert process.state == ProcessState.PAUSED

        # Resume
        process = await engine.resume_process(process.id)
        assert process.state == ProcessState.RUNNING

    @pytest.mark.asyncio
    async def test_get_process_status(self, engine, store, registry):
        """Test getting detailed process status."""
        definition = ProcessDefinition(
            id="status_test",
            name="Status Test",
            first_task_id="task1",
            tasks={
                "task1": TaskDefinition(id="task1", name="Task 1", auto=False),
            },
        )

        process = await engine.create_process(definition)
        await engine.start_process(process.id)

        status = await engine.get_process_status(process.id)

        assert status["process"]["id"] == process.id
        assert status["process"]["definition"] == "Status Test"
        assert status["process"]["state"] == "running"
        assert len(status["tasks"]) == 1


class TestConventionBasedHumanTask:
    """Test the convention-based auto:false pattern for human tasks.

    Human tasks use auto:false with form schema in task properties.
    External callers read task properties to render UI, then call
    engine.complete_task() with user data to resume the workflow.
    """

    @pytest.mark.asyncio
    async def test_human_task_with_form_schema(self, engine, store):
        """Test full lifecycle: task properties as form schema, external completion,
        and result accessible to downstream tasks."""
        definition = ProcessDefinition(
            id="human_task_test",
            name="Human Task Test",
            first_task_id="get_input",
            tasks={
                "get_input": TaskDefinition(
                    id="get_input",
                    name="Get User Input",
                    auto=False,
                    properties={
                        "type": "text_input",
                        "prompt": "Enter your name",
                        "required": True,
                    },
                ),
                "process_input": TaskDefinition(
                    id="process_input",
                    name="Process Input",
                    action="counting",
                ),
            },
            routings=[
                RoutingDefinition(
                    id="r1", source_task_id="get_input", dest_task_id="process_input"
                ),
            ],
        )

        CountingAction.execution_count = 0

        # Create and start process
        process = await engine.create_process(definition)
        await engine.start_process(process.id)

        # Process should be running, waiting for the manual task
        process = await store.load_process(process.id)
        assert process.state == ProcessState.RUNNING

        # Get pending tasks - should find our human task
        pending = await engine.get_pending_tasks(process.id)
        assert len(pending) == 1
        task = pending[0]
        assert task.task_definition_id == "get_input"
        assert task.state == TaskState.READY

        # External caller reads task definition properties (form schema)
        task_def = definition.tasks["get_input"]
        assert task_def.properties["type"] == "text_input"
        assert task_def.properties["prompt"] == "Enter your name"
        assert task_def.properties["required"] is True

        # External caller completes task with user data
        user_response = "Alice"
        await engine.complete_task(task.id, TaskResult.ok(output=user_response))

        # Result should be stored in process properties
        process = await store.load_process(process.id)
        assert process.properties["__task_output_get_input"] == "Alice"

        # Downstream task should have executed
        assert CountingAction.execution_count == 1

        # Process should be complete
        assert process.state == ProcessState.COMPLETE

    @pytest.mark.asyncio
    async def test_human_task_with_structured_output(self, engine, store):
        """Test human task with structured (dict) output from the user."""
        definition = ProcessDefinition(
            id="structured_test",
            name="Structured Human Task",
            first_task_id="form_task",
            tasks={
                "form_task": TaskDefinition(
                    id="form_task",
                    name="Fill Form",
                    auto=False,
                    properties={
                        "type": "form",
                        "fields": [
                            {"name": "first_name", "label": "First Name", "required": True},
                            {"name": "email", "label": "Email", "required": True},
                        ],
                    },
                ),
                "done": TaskDefinition(id="done", name="Done"),
            },
            routings=[
                RoutingDefinition(id="r1", source_task_id="form_task", dest_task_id="done"),
            ],
        )

        process = await engine.create_process(definition)
        await engine.start_process(process.id)

        pending = await engine.get_pending_tasks(process.id)
        assert len(pending) == 1

        # Complete with structured data
        form_data = {"first_name": "Bob", "email": "bob@example.com"}
        await engine.complete_task(pending[0].id, TaskResult.ok(output=form_data))

        # Verify structured data stored in process properties
        process = await store.load_process(process.id)
        assert process.properties["__task_output_form_task"] == form_data
        assert process.state == ProcessState.COMPLETE


class TestTaskSync:
    @pytest.mark.asyncio
    async def test_sync_task_waits_for_all_branches(self, engine, store, registry):
        """Test that sync tasks wait for all parallel branches."""
        definition = ProcessDefinition(
            id="sync_test",
            name="Sync Test",
            first_task_id="start",
            tasks={
                "start": TaskDefinition(id="start", name="Start"),
                "branch_a": TaskDefinition(id="branch_a", name="Branch A", auto=False),
                "branch_b": TaskDefinition(id="branch_b", name="Branch B", auto=False),
                "join": TaskDefinition(id="join", name="Join", synchronized=True),
            },
            routings=[
                RoutingDefinition(
                    id="r1", source_task_id="start", dest_task_id="branch_a", parallel=True
                ),
                RoutingDefinition(
                    id="r2", source_task_id="start", dest_task_id="branch_b", parallel=True
                ),
                RoutingDefinition(id="r3", source_task_id="branch_a", dest_task_id="join"),
                RoutingDefinition(id="r4", source_task_id="branch_b", dest_task_id="join"),
            ],
        )

        process = await engine.create_process(definition)
        await engine.start_process(process.id)

        # Should have two pending manual tasks
        pending = await engine.get_pending_tasks(process.id)
        task_ids = {t.task_definition_id for t in pending}
        assert task_ids == {"branch_a", "branch_b"}

        # Complete branch_a
        branch_a = next(t for t in pending if t.task_definition_id == "branch_a")
        await engine.complete_task(branch_a.id, TaskResult.ok())

        # Join should still be waiting (branch_b not done)
        process = await store.load_process(process.id)
        assert process.state == ProcessState.RUNNING

        # Complete branch_b
        pending = await engine.get_pending_tasks(process.id)
        branch_b = next(t for t in pending if t.task_definition_id == "branch_b")
        await engine.complete_task(branch_b.id, TaskResult.ok())

        # Now process should be complete
        process = await store.load_process(process.id)
        assert process.state == ProcessState.COMPLETE


# ===========================================================================
# fail_process() Tests
# ===========================================================================


def _human_task_definition():
    """A workflow with a human task that stays in READY state."""
    return ProcessDefinition(
        id="human-wf",
        name="Human Workflow",
        version=1,
        first_task_id="manual",
        tasks={
            "manual": TaskDefinition(id="manual", name="Manual Step", auto=False),
            "done": TaskDefinition(id="done", name="Done", auto=True),
        },
        routings=[
            RoutingDefinition(id="r1", source_task_id="manual", dest_task_id="done"),
        ],
    )


class TestFailProcess:
    """Tests for WorkflowEngine.fail_process()."""

    async def test_fail_running_process(self, engine, store):
        """A running process can be moved to FAILED state."""
        defn = _human_task_definition()
        process = await engine.create_process(defn)
        await engine.start_process(process.id)

        result = await engine.fail_process(process.id, "Timed out")

        assert result.state == ProcessState.FAILED
        assert result.properties["__error__"] == "Timed out"
        assert result.completed_at is not None

        # Verify persisted state
        reloaded = await store.load_process(process.id)
        assert reloaded.state == ProcessState.FAILED

    async def test_fail_paused_process(self, engine, store):
        """A paused process can also be failed."""
        defn = _human_task_definition()
        process = await engine.create_process(defn)
        await engine.start_process(process.id)
        await engine.pause_process(process.id)

        result = await engine.fail_process(process.id, "Cancelled")
        assert result.state == ProcessState.FAILED

    async def test_fail_created_process(self, engine, store):
        """A CREATED (not yet started) process can be failed."""
        defn = _human_task_definition()
        process = await engine.create_process(defn)

        result = await engine.fail_process(process.id, "No longer needed")
        assert result.state == ProcessState.FAILED

    async def test_fail_completed_process_raises(self, engine, simple_definition):
        """Cannot fail a process that is already COMPLETE."""
        process = await engine.create_process(simple_definition)
        await engine.start_process(process.id)

        with pytest.raises(InvalidStateTransitionError, match="terminal state"):
            await engine.fail_process(process.id, "Too late")

    async def test_fail_already_failed_process_raises(self, engine, store):
        """Cannot fail a process that is already FAILED."""
        defn = _human_task_definition()
        process = await engine.create_process(defn)
        await engine.start_process(process.id)
        await engine.fail_process(process.id, "First failure")

        with pytest.raises(InvalidStateTransitionError, match="terminal state"):
            await engine.fail_process(process.id, "Second failure")

    async def test_fail_sets_tasks_to_failed(self, engine, store):
        """All non-terminal tasks are moved to FAILED when process is failed."""
        defn = _human_task_definition()
        process = await engine.create_process(defn)
        await engine.start_process(process.id)

        # Verify task is READY before failing
        pending = await engine.get_pending_tasks(process.id)
        assert len(pending) == 1
        assert pending[0].state == TaskState.READY

        await engine.fail_process(process.id, "Cleanup")

        # Check the task is now FAILED
        tasks = await store.load_tasks_for_process(process.id)
        for task in tasks:
            assert task.state == TaskState.FAILED
            assert task.error == "Cleanup"
            assert task.completed_at is not None

    async def test_fail_process_not_found_raises(self, engine):
        """Failing a non-existent process raises ProcessNotFoundError."""
        with pytest.raises(ProcessNotFoundError):
            await engine.fail_process("nonexistent-id", "reason")

File: zebra-rs/tests/test_engine.rs
//! Integration tests for the workflow engine.

use std::collections::HashMap;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;

use async_trait::async_trait;
use serde_json::json;

use zebra::core::errors::ExecutionResult;
use zebra::core::models::{
    ProcessDefinition, ProcessState, RoutingDefinition, TaskDefinition, TaskInstance, TaskResult,
};
use zebra::core::WorkflowEngine;
use zebra::storage::InMemoryStore;
use zebra::tasks::{ActionRegistry, ExecutionContext, TaskAction};
use zebra::StateStore;

// ============================================================================
// Test Actions
// ============================================================================

/// A test action that counts executions.
struct CountingAction {
    counter: Arc<AtomicUsize>,
}

impl CountingAction {
    fn new(counter: Arc<AtomicUsize>) -> Self {
        Self { counter }
    }
}

#[async_trait]
impl TaskAction<InMemoryStore> for CountingAction {
    async fn run(
        &self,
        _task: &TaskInstance,
        _context: &ExecutionContext<InMemoryStore>,
    ) -> ExecutionResult<TaskResult> {
        let count = self.counter.fetch_add(1, Ordering::SeqCst) + 1;
        Ok(TaskResult::ok_with_output(json!({"count": count})))
    }
}

/// A test action that always fails.
struct FailingAction;

#[async_trait]
impl TaskAction<InMemoryStore> for FailingAction {
    async fn run(
        &self,
        _task: &TaskInstance,
        _context: &ExecutionContext<InMemoryStore>,
    ) -> ExecutionResult<TaskResult> {
        Ok(TaskResult::fail("Intentional failure"))
    }
}

// ============================================================================
// Test Fixtures
// ============================================================================

fn create_simple_definition() -> ProcessDefinition {
    let mut tasks = HashMap::new();
    tasks.insert(
        "start".to_string(),
        TaskDefinition::new("start", "Start").with_action("counting"),
    );
    tasks.insert(
        "end".to_string(),
        TaskDefinition::new("end", "End").with_action("counting"),
    );

    ProcessDefinition::new("simple", "Simple Workflow", "start", tasks).with_routings(vec![
        RoutingDefinition::new("r1", "start", "end"),
    ])
}

fn create_parallel_definition() -> ProcessDefinition {
    let mut tasks = HashMap::new();
    tasks.insert("start".to_string(), TaskDefinition::new("start", "Start"));
    tasks.insert(
        "branch_a".to_string(),
        TaskDefinition::new("branch_a", "Branch A").with_action("counting"),
    );
    tasks.insert(
        "branch_b".to_string(),
        TaskDefinition::new("branch_b", "Branch B").with_action("counting"),
    );
    tasks.insert(
        "join".to_string(),
        TaskDefinition::new("join", "Join")
            .with_synchronized(true)
            .with_action("counting"),
    );

    ProcessDefinition::new("parallel", "Parallel Workflow", "start", tasks).with_routings(vec![
        RoutingDefinition::new("r1", "start", "branch_a").with_parallel(true),
        RoutingDefinition::new("r2", "start", "branch_b").with_parallel(true),
        RoutingDefinition::new("r3", "branch_a", "join"),
        RoutingDefinition::new("r4", "branch_b", "join"),
    ])
}

fn create_engine(counter: Arc<AtomicUsize>) -> WorkflowEngine<InMemoryStore> {
    let store = Arc::new(InMemoryStore::new());
    let mut registry = ActionRegistry::new();

    let counter_clone = counter.clone();
    registry.register_action("counting", move || {
        Arc::new(CountingAction::new(counter_clone.clone()))
    });
    registry.register_action("failing", || Arc::new(FailingAction));

    WorkflowEngine::new(store, Arc::new(registry))
}

// ============================================================================
// Tests
// ============================================================================

#[tokio::test]
async fn test_create_process() {
    let counter = Arc::new(AtomicUsize::new(0));
    let engine = create_engine(counter);
    let definition = create_simple_definition();

    let process = engine
        .create_process(definition.clone(), None, None, None)
        .await
        .unwrap();

    assert_eq!(process.state, ProcessState::Created);
    assert_eq!(process.definition_id, definition.id);

    // Definition should be saved
    let loaded = engine.store.load_definition(&definition.id).await.unwrap();
    assert!(loaded.is_some());
}

#[tokio::test]
async fn test_start_process() {
    let counter = Arc::new(AtomicUsize::new(0));
    let engine = create_engine(counter.clone());
    let definition = create_simple_definition();

    let process = engine
        .create_process(definition, None, None, None)
        .await
        .unwrap();

    let process = engine.start_process(&process.id).await.unwrap();

    assert_eq!(process.state, ProcessState::Complete);
    // Both tasks should have run
    assert_eq!(counter.load(Ordering::SeqCst), 2);
}

#[tokio::test]
async fn test_process_with_properties() {
    let counter = Arc::new(AtomicUsize::new(0));
    let engine = create_engine(counter);
    let definition = create_simple_definition();

    let mut props = HashMap::new();
    props.insert("key".to_string(), json!("value"));

    let process = engine
        .create_process(definition, Some(props), None, None)
        .await
        .unwrap();

    assert_eq!(
        process.properties.get("key").unwrap(),
        &json!("value")
    );
}

#[tokio::test]
async fn test_manual_task() {
    let counter = Arc::new(AtomicUsize::new(0));
    let engine = create_engine(counter);

    let mut tasks = HashMap::new();
    tasks.insert(
        "manual_task".to_string(),
        TaskDefinition::new("manual_task", "Manual Task").with_auto(false),
    );
    tasks.insert("end".to_string(), TaskDefinition::new("end", "End"));

    let definition = ProcessDefinition::new("manual", "Manual Workflow", "manual_task", tasks)
        .with_routings(vec![RoutingDefinition::new("r1", "manual_task", "end")]);

    let process = engine
        .create_process(definition, None, None, None)
        .await
        .unwrap();
    engine.start_process(&process.id).await.unwrap();

    // Process should be running, waiting for manual task
    let process = engine.store.load_process(&process.id).await.unwrap().unwrap();
    assert_eq!(process.state, ProcessState::Running);

    // Get pending tasks
    let pending = engine.get_pending_tasks(&process.id).await.unwrap();
    assert_eq!(pending.len(), 1);
    assert_eq!(pending[0].task_definition_id, "manual_task");

    // Complete the manual task
    engine
        .complete_task(&pending[0].id, Some(TaskResult::ok_with_output(json!("done"))))
        .await
        .unwrap();

    // Process should be complete now
    let process = engine.store.load_process(&process.id).await.unwrap().unwrap();
    assert_eq!(process.state, ProcessState::Complete);
}

#[tokio::test]
async fn test_parallel_execution() {
    let counter = Arc::new(AtomicUsize::new(0));
    let engine = create_engine(counter.clone());
    let definition = create_parallel_definition();

    let process = engine
        .create_process(definition, None, None, None)
        .await
        .unwrap();
    engine.start_process(&process.id).await.unwrap();

    let process = engine.store.load_process(&process.id).await.unwrap().unwrap();
    assert_eq!(process.state, ProcessState::Complete);

    // Should have executed: branch_a, branch_b, join
    assert_eq!(counter.load(Ordering::SeqCst), 3);
}

#[tokio::test]
async fn test_pause_resume() {
    let counter = Arc::new(AtomicUsize::new(0));
    let engine = create_engine(counter);

    let mut tasks = HashMap::new();
    tasks.insert(
        "task1".to_string(),
        TaskDefinition::new("task1", "Task 1").with_auto(false),
    );
    tasks.insert("task2".to_string(), TaskDefinition::new("task2", "Task 2"));

    let definition = ProcessDefinition::new("pausable", "Pausable", "task1", tasks)
        .with_routings(vec![RoutingDefinition::new("r1", "task1", "task2")]);

    let process = engine
        .create_process(definition, None, None, None)
        .await
        .unwrap();
    engine.start_process(&process.id).await.unwrap();

    // Pause
    let process = engine.pause_process(&process.id).await.unwrap();
    assert_eq!(process.state, ProcessState::Paused);

    // Resume
    let process = engine.resume_process(&process.id).await.unwrap();
    assert_eq!(process.state, ProcessState::Running);
}

#[tokio::test]
async fn test_get_process_status() {
    let counter = Arc::new(AtomicUsize::new(0));
    let engine = create_engine(counter);

    let mut tasks = HashMap::new();
    tasks.insert(
        "task1".to_string(),
        TaskDefinition::new("task1", "Task 1").with_auto(false),
    );

    let definition = ProcessDefinition::new("status_test", "Status Test", "task1", tasks);

    let process = engine
        .create_process(definition, None, None, None)
        .await
        .unwrap();
    engine.start_process(&process.id).await.unwrap();

    let status = engine.get_process_status(&process.id).await.unwrap();

    assert_eq!(status.id, process.id);
    assert_eq!(status.definition_name, "Status Test");
    assert_eq!(status.state, ProcessState::Running);
    assert_eq!(status.tasks.len(), 1);
}

#[tokio::test]
async fn test_sync_task_waits_for_all_branches() {
    let counter = Arc::new(AtomicUsize::new(0));
    let engine = create_engine(counter);

    let mut tasks = HashMap::new();
    tasks.insert("start".to_string(), TaskDefinition::new("start", "Start"));
    tasks.insert(
        "branch_a".to_string(),
        TaskDefinition::new("branch_a", "Branch A").with_auto(false),
    );
    tasks.insert(
        "branch_b".to_string(),
        TaskDefinition::new("branch_b", "Branch B").with_auto(false),
    );
    tasks.insert(
        "join".to_string(),
        TaskDefinition::new("join", "Join").with_synchronized(true),
    );

    let definition = ProcessDefinition::new("sync_test", "Sync Test", "start", tasks).with_routings(
        vec![
            RoutingDefinition::new("r1", "start", "branch_a").with_parallel(true),
            RoutingDefinition::new("r2", "start", "branch_b").with_parallel(true),
            RoutingDefinition::new("r3", "branch_a", "join"),
            RoutingDefinition::new("r4", "branch_b", "join"),
        ],
    );

    let process = engine
        .create_process(definition, None, None, None)
        .await
        .unwrap();
    engine.start_process(&process.id).await.unwrap();

    // Should have two pending manual tasks
    let pending = engine.get_pending_tasks(&process.id).await.unwrap();
    let task_ids: std::collections::HashSet<_> = pending
        .iter()
        .map(|t| t.task_definition_id.clone())
        .collect();
    assert!(task_ids.contains("branch_a"));
    assert!(task_ids.contains("branch_b"));

    // Complete branch_a
    let branch_a = pending
        .iter()
        .find(|t| t.task_definition_id == "branch_a")
        .unwrap();
    engine
        .complete_task(&branch_a.id, Some(TaskResult::ok()))
        .await
        .unwrap();

    // Join should still be waiting (branch_b not done)
    let process = engine.store.load_process(&process.id).await.unwrap().unwrap();
    assert_eq!(process.state, ProcessState::Running);

    // Complete branch_b
    let pending = engine.get_pending_tasks(&process.id).await.unwrap();
    let branch_b = pending
        .iter()
        .find(|t| t.task_definition_id == "branch_b")
        .unwrap();
    engine
        .complete_task(&branch_b.id, Some(TaskResult::ok()))
        .await
        .unwrap();

    // Now process should be complete
    let process = engine.store.load_process(&process.id).await.unwrap().unwrap();
    assert_eq!(process.state, ProcessState::Complete);
}

#[tokio::test]
async fn test_task_failure() {
    let counter = Arc::new(AtomicUsize::new(0));
    let engine = create_engine(counter);

    let mut tasks = HashMap::new();
    tasks.insert(
        "failing_task".to_string(),
        TaskDefinition::new("failing_task", "Failing Task").with_action("failing"),
    );
    tasks.insert("end".to_string(), TaskDefinition::new("end", "End"));

    let definition = ProcessDefinition::new("failing", "Failing Workflow", "failing_task", tasks)
        .with_routings(vec![RoutingDefinition::new("r1", "failing_task", "end")]);

    let process = engine
        .create_process(definition, None, None, None)
        .await
        .unwrap();

    // When a task action fails, it doesn't route to the next task.
    // The process still completes (though ideally it should be marked Failed).
    let result = engine.start_process(&process.id).await;
    assert!(result.is_ok());

    // Verify the process completed but the end task was never reached
    let status = engine.get_process_status(&process.id).await.unwrap();
    assert_eq!(status.state, ProcessState::Complete);
    // The failing task should still be in the task list with Failed state
    // (Note: in current impl it may have been cleaned up, so we just verify process ended)
}
